src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Data/BindingListCollectionView.cs

```diff
@@ -2009,6 +2009,18 @@ private void OnListChanged(object sender, ListChangedEventArgs args)
                 break;
 
             case ListChangedType.ItemChanged:
+                // if there is no PropertyDescriptor, then ItemChanged refers to a Replace event
+                // (IBindingList indexer set) and not a property change of an item
+                if (args.PropertyDescriptor == null)
```

Comment:
But there is one more thing that we can check here, when ItemChanged is raised from SetItem method, we use this constructor `ListChangedEventArgs(ListChangedType listChangedType, int newIndex) : this(listChangedType, newIndex, -1)` [link](https://github.com/dotnet/runtime/blob/1da23b146496d95f017439a48f62e78ec15b0289/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ListChangedEventArgs.cs#L11-L13) which leaves `OldIndex = -1`, however, when it raises ItemChanged for property changed event it calls `ListChangedEventArgs(ListChangedType listChangedType, int newIndex, PropertyDescriptor? propDesc)` [link](https://github.com/dotnet/runtime/blob/1da23b146496d95f017439a48f62e78ec15b0289/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ListChangedEventArgs.cs#L15-L19), which sets `OldIndex = NewIndex`.

I think we can add that check here to prevent raising the event twice.

Reply:
@dipeshmsft 

> while going through how BindingList raises ItemChanged event, I found [this](https://github.com/dotnet/runtime/blob/1da23b146496d95f017439a48f62e78ec15b0289/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/BindingList.cs#L524-L528) which suggests that PropertyDescriptor can be null even in property change event.

Good observation. I think this is a fallback behavior because in that case we definitely have a [non-null property name](https://github.com/dotnet/runtime/blob/1da23b146496d95f017439a48f62e78ec15b0289/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/BindingList.cs#L474); still, there was no matching property descriptor, which is kind of illegal (the event source was lying about the property name).

> But there is one more thing that we can check here, when ItemChanged is raised from SetItem method, we use this constructor [...] which leaves `OldIndex = -1`,

I actually hesitate to change the condition to `if (args.PropertyDescriptor == null && args.OldIndex < 0)` because it would be very `BindingList<T>` specific. What if other `IBindingList` implementations use other constructors for item replace (or if `BindingList<T>` will be refactored)?

> I think we can add that check here to prevent raising the event twice.

I didn't observe any chances for duplicate events. If I'm not mistaken there will always be just one event, it's just its action type that will be different. For an `ListChangedType.ItemChanged` event now there are 3 possible outcomes (as opposed to the original 2 ones):

1. _We have a property descriptor and `T` implements `INotifyPropertyChanged`_ (original behavior): `ItemChanged` is ignored because the property will be updated in the view anyway (no collection event is raised, there is just the `PropertyChanged` from the item itself)
2. _We have a property descriptor and `T` does **not** implement `INotifyPropertyChanged`_ (original behavior): A Reset action is reported to the view to be able to reflect the property change. The selection jumps to the top of the DataGrid/ListBox or whatever view.
3. _There is no property descriptor_ (the newly added behavior): We treat it as a Replace event. The view will refresh the affected row, the selection remains where it was.

**Remarks**
* I actually think that there should be a `Replace` also in 2. instead of a `Reset` but **that would be a breaking change** (unless we treat it as bugfix because the jumping selection can be really annoying in case of thousands of rows). That reset meant to be the fallback if the UI cannot receive notifications about the changed property but it would be enough to update just the affected row instead of resetting the whole binding. It's just this issue hardly ever bothered anyone because in WPF almost everyone uses `ObservableCollection<T>` instead of `BindingList<T>`. So `BindingListCollectionView` is mainly used only for `DataTable`, which isn't affected by the issue because `DataView` invokes the `ItemChanged` event only for cell updates (ie. the descriptor is never null) and `DataRowView` implements `INotifyPropertyChanged`.
* Now if we don't apply `&& args.OldIndex < 0` to 3. we also have a "breaking change" for the aforementioned semi-illegal case when we had a property name without a matching descriptor. But I think this is still better than tailoring the the behavior for the implementation of `BindingList<T>`. And as I mentioned above updating just the affected row is still better than resetting the whole view, it's just I didn't dare to "fix" that part for the non-exceptional case.

What do you (and/or the others) think? Of course, the final decision is yours, I was just sharing my thoughts. Feel free to apply any changes to my PR.

Reply:
> I actually hesitate to change the condition to if (args.PropertyDescriptor == null && args.OldIndex < 0) because it would be very BindingList<T> specific. What if other IBindingList implementations use other constructors for item replace (or if BindingList<T> will be refactored)?

I agree that using args.OldIndex < 0 will be very BindingList specific and we need some other way to distinguish the cases.

> I didn't observe any chances for duplicate events.

Sorry for the confusion. I was trying to say that, in the current implementation if __T implements `INotifyPropertyChanged`__  and due to `PropertyChanged` we ignore the `ItemChanged` event. However, now there is a possibility that we raise the `CollectionChanged` as a result of this addition, which may affect the other implementations of `IBindingList`.

> There is no property descriptor (the newly added behavior): We treat it as a Replace event. The view will refresh the affected row, the selection remains where it was.

This in itself is the right approach, but this will make it difficult to deal with the semi-illegal cases where `PropertyDescriptor` is null.

Reply:
> Sorry for the confusion. I was trying to say that, in the current implementation if T implements INotifyPropertyChanged and due to PropertyChanged we ignore the ItemChanged event. However, now there is a possibility that we raise the CollectionChanged as a result of this addition, which may affect the other implementations of IBindingList.

For this, one approach can be to check the equality of items from `_cachedList` and `InternalList`. What are your thoughts on this ?

Reply:
> Sorry for the confusion. I was trying to say that, in the current implementation if `T` implements `INotifyPropertyChanged` and due to `PropertyChanged` we ignore the `ItemChanged` event. However, now there is a possibility that we raise the `CollectionChanged` as a result of this addition, which may affect the other implementations of IBindingList.

I don't really see any issue here. Let's check every sub-case separately:
1. `T` implements `INotifyPropertyChanged`: the current code ignores `CollectionChange` supposing it isn't required: `ItemChanged` is observable for the changed property.
   * a.) Property descriptor is not `null`:  the event is ignored, we rely on `INotifyPropertyChanged.PropertyChanged`. **The original behavior remains**.
   * b.) Property descriptor is `null`: In this case `CollectionChange` was simply ignored. Acceptable for the semi-illegal case where we just didn't find a matching property but clearly a bug for an item replace as we fail to update the changed row. I think this case was simply not considered because it never occurs for a `DataTable`: it removes and adds whole rows, never replaces them, and uses `ItemChanged` for property changes (a single cell) only. For the semi-illegal case (when there was a nonexistent property name) this is also a change: we now update the whole row instead of ignoring the event. In worst case we did a refresh for nothing, but if the property name was just misspelled, the update will reveal the change somewhere in the row.
2. `T` does **not** implement `INotifyPropertyChanged`:
   * a.) Property descriptor is not `null`: **The original behavior remains**, ie. the whole binding is reset. But TBH I think this is an overkill, a `Replace` would be better here, too. I think the intention was just to update the whole row along with the changed property. But now the whole list is refreshed (even though we know the index of the changed row), which causes the selection to jump to the first row.
   * b.) Property descriptor is not `null`: We apply the new `Replace` event instead of a `Reset`. As described above, never occurs for a `DataTable`, so the initiator is likely another `IBindingList` with an indexer set, maybe a `BindingList<T>`. Or, the semi-illegal case with no matching property. In which case we have the row index but could not identify the property so again, the `Replace` for the whole row instead the old `Reset` seems just fine.

**Verdict:** The only case where we have a new event that was suppressed before is 1.b.) - but this is the very point of this PR. At least, I cannot imagine a use case where a special `IBindingList` implementation raises events that are better to be suppressed in this case.

Reply:
Your explanation is to the point. The case 2.b fixes a wrong behavior and I am in for that. I am just a bit concerned about the `refresh for nothing` case because that results in 1 more extra event for every item replaced.

Said that I am adding some tests in our test repo to validate the changes and prevent future regressions. And then we will take this PR for a test pass.

Reply:
@koszeggy , while going through other ListChangedType cases, I see that we check the items against `CurrentEditItem` and `CurrentAddItem`. Don't we need to handle those cases for `ItemsChanged` case ?

Reply:
It seems those belong to `IEditableCollectionView` and are related also to `IBindingList.AddNew`. Now I checked what happens in other collections when a new item is added or edited, and when are `EndEdit` or `EndAddNew` (-like) methods called. It seems that they are mainly checked/adjusted when new items are added, an item is removed or when the list is cleared. Specifically:
* `BindingList<T>`: It calls `EndNew` from [ClearItems, InsertItem and RemoveItem](https://github.com/dotnet/runtime/blob/e4fbdb907bb187d7b5ba0668a84347c1058e3219/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/BindingList.cs#L181-L229) but not from [SetItem](https://github.com/dotnet/runtime/blob/e4fbdb907bb187d7b5ba0668a84347c1058e3219/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/BindingList.cs#L233-L248)
* `BindingListCollectionView`: Handles `CurrentAddItem` (and calls `EndAddNew`) for [`ItemDeleted`](https://github.com/dotnet/wpf/blob/984f37ed482d55bc47f2cef29423275d5bc209ee/src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Data/BindingListCollectionView.cs#L1962) but only has an assert for [`ItemAdded`](https://github.com/dotnet/wpf/blob/984f37ed482d55bc47f2cef29423275d5bc209ee/src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Data/BindingListCollectionView.cs#L1928) and does not call `EndAddNew`. Like `BingingList<T>`, it does't check the newly added or edited item for `ItemChange`
* I quickly checked also my [`FastBindingList<T>`](https://docs.kgysoft.net/corelibraries/html/T_KGySoft_ComponentModel_FastBindingList_1.htm) and [`ObservableBindingList<T>`](https://docs.kgysoft.net/corelibraries/html/T_KGySoft_ComponentModel_ObservableBindingList_1.htm) classes (which helped me to discover this bug). The former is completely based on `BindingList<T>` so behaves the same way, whereas the latter is a wrapper class (a bit like `BindingListCollectionView`), and just like the others, [handles only](https://github.com/koszeggy/KGySoft.CoreLibraries/blob/adace160db20afe54c4668ef40336b9a2454b209/KGySoft.CoreLibraries/ComponentModel/_Collections/ObservableBindingList.cs#L1166) the add/remove/clear cases.

**BUT:** The `EndNew`/`EndAddNew` methods (which are are usually [`ICancelAddNew`](https://learn.microsoft.com/en-us/dotnet/api/system.componentmodel.icanceladdnew.endnew) implementations) work with the _index_ of the newly added item so it's kind of obvious that they can ignore the `Replace` event. But `BindingListCollectionView` apparetly combines the new/edited index tracking with `IEditableCollectionView.CurrentAddItem` and `CurrentEditItem`, which require also a _reference_ for the newly added and currently edited items. So I think you are right, they should be checked, too. I think this is then a potential bug that exists in `BindingListCollectionView` independently from the notification issue I tried to address.

However, I don't know how `IEditableCollectionView` implementation should be handled conventionally in such cases as this is literally the first time I investigate this interface. I can't even tell whether maybe an assert would be enough (like for `ItemAdded`) or should we call `EndAddNew`/`CancelEdit` (like in `ItemDeleted`), but I suspect the latter. I think this part should be reviewed/answered by a genuine WPF guy (which I'm not) who can either handle it here or can maybe create a follow-up issue.